Use relative pointers for module initializers section

jkotas opened the pull request:
Fixes #1704

jkotas committed:
Use relative pointers for module initializers section

Fixes #1704

jkotas commented on src/coreclr/nativeaot/Common/src/Internal/Runtime/CompilerHelpers/StartupCodeHelpers.cs:
```diff
@@ -218,12 +218,25 @@ internal static unsafe void RunModuleInitializers()
-                        (*currentCctor++)();
+                        void* cctor;
+
+                        if (Internal.Runtime.MethodTable.SupportsRelativePointers)
```
An alternative solution would be to make the section writeable on non-Windows. (The sections with absolute pointers have to be writeable on non-Windows.)

MichalStrehovsky reviewed (approved):
It would be nice if this was consistent with how we do eager constructors (they're all pointer sized) but since we already have this code, LGTM.

MichalStrehovsky commented:
> It would be nice if this was consistent with how we do eager constructors (they're all pointer sized) but since we already have this code, LGTM.

Oh, looks like we just put those into Data section no matter what. It probably doesn't matter much because they're never looked at once we're done starting up.

jkotas committed:
Feedback, additional cleanup
- Switch back to pointers, change the section to be writeable instead
- Share code to invoked initializers between eager constructors and module initializers
- Make ExternalReferencesTableNode readonly where possible
- Delete unused ThreadingHelpers.cs
